feat: manual alert handling

mataa019 opened the pull request:
Manual Alerts - Feature Implementation

This pull request covers the initial implementation of core functionality for the Manual Alerts module. The following features have been completed:

Completed Tasks:
View Alert Table – Added logic to retrieve and list stored alerts.

Open and Review an Alert – Implemented functionality to fetch and return detailed alert data by ID.

Convert Alert to Case – Added the ability to convert an alert into a case with appropriate data transformation and persistence.

Close Alert as False Positive – Enabled marking and closing of alerts as false positives.

Ingest Alert via Tazama NATS – Integrated with the Tazama NATS broker to consume and process alert messages in real time.

Extract Tenant ID and Transaction Type – Implemented parsing and validation of tenantId and txTp from the transaction object in alert payloads.

Testing and Validation:
Unit and integration tests implemented for core services and processing logic.

Functional validation performed on  Manual alert processing workflows.
Tasks
[1. Manual Alerts - View Alert Table]
[2. Manual Alerts - Open and Review an Alert]
[3. Manual Alerts - Convert Alert to Case]
[4. Manual Alerts - Close Alert as False Positive]
[5. Manual Alerts - Ingest Alert via Tazama NATS
[6. Manual Alerts - Extract Tenant ID and Transaction Type from Transaction Object]

Justus-at-Tazama committed:
Initial commit

mataa019 committed:
feat(core): init NestJS with triage mock API

mataa019 committed:
feat(core): init NestJS with triage mock API

mataa019 committed:
feat(core): init NestJS with triage mock API

mataa019 committed:
feat(core): init NestJS with triage mock API

mataa019 committed:
feat:unit testing for triage auto

mataa019 committed:
feat:implementing the auth service

mataa019 committed:
feat: prisma

Akku-Rawat committed:
feat: audit log

mataa019 committed:
feat/prisma

Akku-Rawat committed:
fix: remove invalid AuditLogController import

Akku-Rawat committed:
feat(prisma): complete initial schema setup

Akku-Rawat committed:
feat: prisma complete initial schema setup

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:auth

mataa019 committed:
feat:Authentication & Authorization

mataa019 committed:
feat:Authentication & Authorization

mataa019 committed:
feat:Authentication & Authorization

mataa019 committed:
feat:Authentication & Authorization

mataa019 committed:
feat:Authentication & Authorization

mataa019 committed:
feat:Authentication & Authorization

mataa019 committed:
Merge pull request #1 from tazama-lf/feature-auth

Feature auth

Akku-Rawat committed:
feat:triage  manual alert triage

Akku-Rawat committed:
feat(triage) :  manual alert triage

mataa019 committed:
Merge branch 'feat/triage-auto-trigger' into feature-auth

mataa019 committed:
Merge pull request #2 from tazama-lf/feature-auth

Feature auth

mataa019 committed:
feat:fixing the conflic merge

mataa019 committed:
feat:fixing the conflic merge dev

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service dev

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat: Auth adding roles decorators

mataa019 committed:
feat: Auth adding roles decorators dev

mataa019 committed:
feat: Unit Test for Auth and Audit

mataa019 committed:
feat: Unit Test for Auth and Audit

mataa019 committed:
feat: Test Coverage

mataa019 committed:
feat: Test Coverage dev

mataa019 committed:
feat: Test Coverage

mataa019 committed:
feat: Test Coverage

mataa019 committed:
feat:Test to Triage Module

mataa019 committed:
feat:Test to Triage Module

mataa019 committed:
feat: Test Coverage for Triage Module

mataa019 committed:
feat: Test Coverage for Triage Module

Akku-Rawat committed:
feat: triage send alert for manual investigation

Akku-Rawat committed:
feat(triage): send alert for manual investigation

mataa019 committed:
feat: test

mataa019 committed:
feat: test

Akku-Rawat committed:
feat(triage): enforce tenant isolation for alert operations

- Restricted update, manual close, and investigation actions to the tenant that owns the alert.
- Prevents unauthorized access or manipulation of alerts across tenants.
- Throws NotFoundException when tenant mismatch occurs.

Akku-Rawat committed:
feat(triage): enforce tenant isolation for alert operations

- Restricted update, manual close, and investigation actions to the tenant that owns the alert.
- Prevents unauthorized access or manipulation of alerts across tenants.
- Throws NotFoundException when tenant mismatch occurs.

mataa019 committed:
feat: add the pre-push file

mataa019 committed:
feat: add the pre-push file

mataa019 committed:
feat: Authentication Guards Added at Controller Level dev

mataa019 committed:
feat: Authentication Guards Added at Controller Level

mataa019 committed:
feat: Authentication Guards Added at Controller Level

mataa019 committed:
feat: Authentication Guards Added at Controller Level

mataa019 committed:
feat: add a .env.template file that can be used via command: cp .env.template .env

mataa019 committed:
feat: add a .env.template file that can be used via command: cp .env.template .env dev

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

scott45 committed:
build: add .github folder in the repo

Justus-at-Tazama committed:
Merge pull request #11 from tazama-lf/config

build: add .github folder in the repo

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
Add .env.example (renamed from .env.template)

mataa019 committed:
Add .env.example (renamed from .env.template)

mataa019 committed:
feat: fixing the esLint and prettier errors

mataa019 committed:
feat: fixing the esLint and prettier errors

mataa019 committed:
Add .env.example (renamed from .env.template)

mataa019 committed:
Add .env.example (renamed from .env.template)

mataa019 committed:
Add .env.example (renamed from .env.template)

mataa019 committed:
Add .env.example (renamed from .env.template)

Akku-Rawat committed:
Merge branch 'Auth-test' into izyane

Akku-Rawat committed:
feat(manual-alert): add alert list and details retrieval

- Implement getAlertsForUser to fetch tenant alerts with filtering (type, priority, status), keyword search, sorting, pagination, and input validation;.
- Update getAlertDetails to retrieve alert details with tenantId access control, system logging for access (userId, timestamp).

Akku-Rawat committed:
feat(manual-alert): add alert list and details retrieval

- Implement getAlertsForUser to fetch tenant alerts with filtering (type, priority, status), keyword search, sorting, pagination, and input validation;.
- Update getAlertDetails to retrieve alert details with tenantId access control, system logging for access (userId, timestamp).

mwan9ilwa committed:
fixed coverage threshold

mwan9ilwa committed:
Test pre-commit hook with user decorator tests dev

mwan9ilwa committed:
Test pre-commit hook with user decorator tests

mwan9ilwa committed:
Update Husky configuration to v10 format

mwan9ilwa committed:
Update Husky configuration to v10 format

Akku-Rawat committed:
Merge branch 'izyane' into Auth-Test

Akku-Rawat committed:
Merge pull request #17 from tazama-lf/Auth-Test

Auth test

Akku-Rawat committed:
refactor: auth in triage controller

Akku-Rawat committed:
refactor: auth in triage controller dev

Akku-Rawat committed:
Merge branch 'izyane' into manual-alert

Akku-Rawat committed:
remove: redundant auth guard dev

Akku-Rawat committed:
remove: redundant auth guard

Akku-Rawat committed:
feat:Manual Alert add close alert in case of false positive

- Create CloseAlertDto to capture the reason for closing an alert.
- Update manualCloseAlert to close alerts with a reason and log the event with reason and timestamp in the audit log.
- Modify updateAlertData to prevent updates on alerts with a closed status.
- Remove redundant AutoCloseAlertDto.

Akku-Rawat committed:
feat(manual-alert): add close alert in case of false positive

- Create CloseAlertDto to capture the reason for closing an alert.
- Update manualCloseAlert to close alerts with a reason and log the event with reason and timestamp in the audit log.
- Modify updateAlertData to prevent updates on alerts with a closed status.
- Remove redundant AutoCloseAlertDto.

mwan9ilwa committed:
fixed all dto test cases

mwan9ilwa committed:
fixed all dto test cases

mwan9ilwa committed:
fixed failing test cases

mwan9ilwa committed:
fixed failing test cases

mwan9ilwa committed:
Fix lint errors in triage test files

- Added ESLint disable directives for test-specific patterns
- Fixed unused variable warning in triage.service.spec.ts
- Maintained test functionality with 244 passing tests
- Achieved 98.15% test coverage for triage module

mwan9ilwa committed:
Fix lint errors in triage test files dev

- Added ESLint disable directives for test-specific patterns
- Fixed unused variable warning in triage.service.spec.ts
- Maintained test functionality with 244 passing tests
- Achieved 98.15% test coverage for triage module

Akku-Rawat committed:
feat:manual alert add convert alert to case

- Add ConvertAlertToCase DTO to capture additional case input
- Implement convertToCase API to create case from alert, mark alert as converted, and log audit trail
- Updated CaseStatus, AlertStatus   and CaseCreationType enum

Akku-Rawat committed:
feat(manual-alert): add convert alert to case

- Add ConvertAlertToCase DTO to capture additional case input
- Implement convertToCase API to create case from alert, mark alert as converted, and log audit trail
- Updated CaseStatus, AlertStatus   and CaseCreationType enum

mwan9ilwa committed:
fixed failing test cases

mwan9ilwa committed:
fixed failing test cases

mwan9ilwa committed:
fixed failing test cases

mwan9ilwa committed:
fixed failing test cases

Akku-Rawat committed:
fix: lint errors

Akku-Rawat committed:
fix: lint errors

mataa019 committed:
feat: Ingest Alert via Tazama NATS

mataa019 committed:
feat: Ingest Alert via Tazama NATS dev

mataa019 committed:
fix lint nats implementation

mataa019 committed:
fix lint

Akku-Rawat committed:
fix: failing test case

-- fixed failing test cases
- removed redundant source and txtp field from request body

Akku-Rawat committed:
fix: failing test case

-- fixed failing test cases
- removed redundant source and txtp field from request body

Akku-Rawat committed:
Merge branch 'manual-alert-handling' of https://github.com/tazama-lf/case-management-system into manual-alert-handling

mataa019 committed:
fix lint nats implementation

mataa019 committed:
fix lint

mwan9ilwa committed:
fix: resolve failing tests and improve AlertMessageDto coverage

- Add @IsOptional decorator to source field in AlertMessageDto.dto.ts
- Fix app.module.spec.ts by adding required STARTUP_TYPE env var and startup factory mock
- Update nats-startup.spec.ts test expectations to match actual implementation
- Enhance AlertMessageDto.dto.spec.ts with comprehensive test coverage:
  * Add 11 new test cases (doubled from 11 to 22 tests)
  * Test all field validation scenarios including type validation
  * Test optional field handling with null/undefined values
  * Test complex nested object validation
  * Test enum validation for Priority and AlertStatus
  * Achieve 100% coverage on AlertMessageDto.dto.ts

Results:
- All 288 tests now passing (was 286 with 2 failures)
- Overall coverage increased to 97.89% statements, 97.72% lines
- Exceeded all coverage thresholds (95%+)
- AlertMessageDto.dto.ts has 100% coverage across all metrics

mwan9ilwa committed:
fix: resolve failing tests and improve AlertMessageDto coverage

- Add @IsOptional decorator to source field in AlertMessageDto.dto.ts
- Fix app.module.spec.ts by adding required STARTUP_TYPE env var and startup factory mock
- Update nats-startup.spec.ts test expectations to match actual implementation
- Enhance AlertMessageDto.dto.spec.ts with comprehensive test coverage:
  * Add 11 new test cases (doubled from 11 to 22 tests)
  * Test all field validation scenarios including type validation
  * Test optional field handling with null/undefined values
  * Test complex nested object validation
  * Test enum validation for Priority and AlertStatus
  * Achieve 100% coverage on AlertMessageDto.dto.ts

Results:
- All 288 tests now passing (was 286 with 2 failures)
- Overall coverage increased to 97.89% statements, 97.72% lines
- Exceeded all coverage thresholds (95%+)
- AlertMessageDto.dto.ts has 100% coverage across all metrics

mataa019 committed:
fixing the lint issues

mataa019 committed:
fixing the lint issues

mataa019 committed:
fixing the lint issues

mataa019 committed:
fixing the lint issues

mataa019 committed:
fix merging

mataa019 committed:
fix lint

mataa019 committed:
Resolve merge conflicts with dev branch

mataa019 committed:
fix:gpg verify

UmairKhan-Paysys commented on .npmrc.example:
Just the .npmrc file should suffice. No need for a .npmrc.example file.

Akku-Rawat replied:
Kept only .npmrc and removed .npmrc.example

UmairKhan-Paysys commented on .npmrc.example:
```diff
@@ -0,0 +1,2 @@
+@tazama-lf:registry=https://npm.pkg.github.com
+//npm.pkg.github.com/:_authToken=token_source
```
You can have the github token saved in you OS environment variable as `GH_TOKEN` which will be accessible here.
```suggestion
//npm.pkg.github.com/:_authToken=${GH_TOKEN}
```

Akku-Rawat replied:
updated .npmrc and added GitHub token reference

UmairKhan-Paysys commented on README.md:
I understand you have made the changes regarding the README.md in the other branch. Pointing it out here that this needs to be merged.

Akku-Rawat replied:
Updated the README in this branch to align with the auth_service branch for an easier merge

UmairKhan-Paysys commented on prisma/schema.prisma:
```diff
@@ -15,7 +13,133 @@ model AuditLog {
-}
\ No newline at end of file
+}
+model Alert {
+  alert_id       String   @id @default(uuid()) @db.Uuid
+  tenant_id      String   @db.Uuid
+  priority       Priority? @default(LOW)
+  source         String?   @db.VarChar(50)
+  txtp           String?   @db.VarChar(50)
+  message        String   @db.VarChar(100)
+  alert_data     Json
+  transaction    Json
+  network_map    Json
+  alert_status   AlertStatus @default(NEW)
+  confidence_per Int
+  created_at     DateTime @default(now()) @db.Timestamp(6)
+  case_id        String?   @db.Uuid
+  case           Case?     @relation(fields: [case_id], references: [case_id])
+  @@map("alerts")
+}
```
I noticed there was an alert_type missing in the ERD which is used to identify whether the alert is `FRAUD` or `AML` or both. Please add it here.

Akku-Rawat replied:
Added an alert_type field to the Alert model and introduced an alertType enum.

UmairKhan-Paysys commented on src/auth/auth.module.ts:
```diff
@@ -4,13 +4,15 @@ import { JwtStrategy } from './jwt.strategy';
+import { PrismaModule } from '../../prisma/prisma.module';
+import { AuditLogModule } from '../audit/auditLog.module';
+
+import { Logger } from '@nestjs/common';
```
Tazama already has a [LoggerService](https://github.com/tazama-lf/docs/tree/dev/Technical/Logging) which is what we should be using for logging. Instead of using the NestJS built-in loggerService.

mataa019 replied:
We are now currently using the  Tazama LoggerService

UmairKhan-Paysys commented on src/main.ts:
```diff
@@ -1,8 +1,22 @@
-  await app.listen(process.env.PORT ?? 3000);
+  const logger = new Logger('Bootstrap');
+  const app = await NestFactory.create(AppModule, {
+    logger: ['log', 'error', 'warn', 'debug', 'verbose'],
```
We should be using the Tazama LoggerService

mataa019 replied:
Have updated the logger to use the  Tazama LoggerService

UmairKhan-Paysys commented on src/triage/triage.service.ts:
```diff
@@ -1,63 +1,482 @@
+    }
+
+    try {
+      const updatedAlert = await this.prisma.alert.update({
+        where: { alert_id: alertId },
+        data: {
+          confidence_per: dto.confidence_per,
+          priority: dto.priority,
+        },
```
We should also be setting an alert_type.

Akku-Rawat replied:
Modified the updateAlert service method to include alertType in the Prisma update call data, also added alertType enum to updateAlertDto

UmairKhan-Paysys commented on src/triage/triage.service.ts:
```diff
@@ -1,63 +1,482 @@
+    tenantId: string,
+  ) {
+    const alert = await this.prisma.alert.findUnique({
+      where: { alert_id: alertId },
```
Fetching using tenantId here might be better instead of having a conditional check later in the code.
```suggestion
      where: { alert_id: alertId, tenantId: tenantId },
```

Akku-Rawat replied:
Updated alert fetching by using both alertId and tenantId directly

UmairKhan-Paysys commented on src/triage/triage.service.ts:
```diff
@@ -1,63 +1,482 @@
+      throw new NotFoundException(
+        `Alert ${alertId} not accessible for this tenant`,
+      );
+    }
```
No need for this here. If we check the tenantId while fetching the alert.

Akku-Rawat replied:
Have updated fetching logic and removed this condition

UmairKhan-Paysys commented on src/triage/triage.service.ts:
```diff
@@ -1,63 +1,482 @@
+    caseType: CaseType,
+    userId: string,
+    tenantId: string,
+  ) {
```
I want to understand why there is a convertToCase function and a investigateAlert. Both seem to be doing the same thing.

Akku-Rawat replied:
Before the manual alert stories were introduced, we had started implementing stories 1d, 1e, and 1f, which involved updating the alertStatus to SENT_FOR_INVESTIGATION and creating a case. These methods were leftovers from that earlier work and should have been removed. I’ve now removed them from this branch,  leaving only manual alert stories.

UmairKhan-Paysys commented on src/triage/triage.service.ts:
```diff
@@ -1,63 +1,482 @@
+    }
+  }
+
+  async convertToCase(
+    alertId: string,
+    convertAlertToCase: ConvertAlertToCase,
+    userId: string,
+    tenantId: string,
+  ) {
```
This and the investigateAlert are same. I suggest we keep one,

Akku-Rawat replied:
Have removed investigateAlert  and kept convertToCase

UmairKhan-Paysys commented on src/triage/triage.controller.ts:
```diff
@@ -1,13 +1,162 @@
+    }
+
+    return alert;
+  }
```
We should be just saving the alert here and not moving it directly to investigation.

Akku-Rawat replied:
Same as before, this was originally based on story 1g, where we sent the alert for investigation if no confidence threshold was present. I’ve now removed the environment variable check and case creation logic from this branch to facilitate an easier review process, leaving only manual alert stories.

UmairKhan-Paysys commented on src/triage/triage.controller.ts:
```diff
@@ -1,13 +1,162 @@
+      convertAlertToCase,
+      userId,
+      tenantId,
+    );
```
This also converts to case which makes is fine but If I'm not mistaken. The investigateAlert also seems to be creating a case instead of holding it for manual triage.

Akku-Rawat replied:
Have removed investigateAlert  and kept convertToCase

UmairKhan-Paysys reviewed (changes requested):
Identified some issues. Requesting changes. What I also need clarity on is that the other PR is now being maintained separately. So can you kindly also mention which one do you expect to merge first as to get minimal merge conflicts. Thank you.

Akku-Rawat committed:
feat(enum): created alertType enum and updated schema prisma

Akku-Rawat committed:
refactor(manual-alert): apply PR feedback to alert update and flow

- Added \lertType\ field to UpdateAlertDto
- Updated updateAlert service to include \lertType\ in Prisma update data
- Modified manualAutoCloseAlert to fetch by alertId and tenantId
- Removed obsolete investigateAlert service and controller method
- Updated submitAlert to only create alerts (removed case creation logic)

Akku-Rawat committed:
chore: update gitignore, add .npmrc, remove example file, and update README

Akku-Rawat committed:
refactor: update alertType enum

mataa019 committed:
change logger to loggerservice tazama

mataa019 committed:
change logger to loggerservice tazama

Akku-Rawat committed:
fix: failing unit test

mataa019 committed:
update the unit test to use loggerservice

mataa019 committed:
update the unit test to use loggerservice

mataa019 committed:
updating the .env example

mataa019 commented:
Okay you can start with the Auth-Service-PR merge it first before manual-alert-handling

UmairKhan-Paysys commented on src/main.ts:
```diff
@@ -1,8 +1,30 @@
 async function bootstrap() {
-  const app = await NestFactory.create(AppModule);
-  await app.listen(process.env.PORT ?? 3000);
+  // Validate and get the processor config for LoggerService
+  const configuration = validateProcessorConfig();
+  const logger = new LoggerService(configuration);
+
+  const app = await NestFactory.create(AppModule, {
+    logger,
+  });
```
NestJS uses a ConfigModule to load environment variables. Here you are passing the loggerService some configuration even before the NestFactory is initialized. Therefore, when I was trying to run it the loggerService seemed to throw and undefined env error. 
```suggestion
  const app = await NestFactory.create(AppModule);
  const logger = app.get(LoggerService);
  app.useLogger(logger);
```

mataa019 replied:
We have updated our NestJS bootstrap process to follow best practices for environment variable loading and logger initialization. The application now initializes the NestFactory first, retrieves the LoggerService from the DI container, and sets it using app.useLogger(logger). This ensures all configuration is loaded before the logger is used, resolving the previous undefined environment error.

UmairKhan-Paysys commented on src/main.ts:
To fix the loggerService issue you might also need to create a logger.module.ts
```typescript
import { Module } from '@nestjs/common';
import { LoggerService } from '@tazama-lf/frms-coe-lib';
import { ConfigService } from '@nestjs/config';

@Module({
  providers: [
    {
      provide: LoggerService,
      useFactory: (configService: ConfigService) => {
        const configuration = {
          nodeEnv: configService.get<string>('NODE_ENV', 'dev'),
          maxCPU: Number(configService.get<string>('MAX_CPU', '1')),
          functionName: configService.get<string>('FUNCTION_NAME', 'case-management-system'),
        };
        return new LoggerService(configuration);
      },
      inject: [ConfigService],
    },
  ],
  exports: [LoggerService],
})
export class LoggerModule {}
```

You can then import this module into any other module to utilize the singleton functionality of NestJS

UmairKhan-Paysys commented on src/nats/nats.module.ts:
```diff
@@ -0,0 +1,11 @@
+  imports: [TriageModule],
+  providers: [NatsStartupService, LoggerService],
+  exports: [NatsStartupService],
+})
```
```suggestion
@Module({
  imports: [TriageModule, LoggerModule],
  providers: [NatsStartupService],
  exports: [NatsStartupService],
})
```
After the LoggerService changes the module files should look something like this.

mataa019 replied:
We have updated NatsModule to import LoggerModule instead of providing logggerservice directly. This ensures proper dependency injection and avoids multiple logger instances.

UmairKhan-Paysys commented on src/nats/nats.startup.ts:
```diff
@@ -0,0 +1,73 @@
+  ) {}
+  private server: IStartupService;
+
+  async onModuleInit() {
+    this.server = new StartupFactory();
+    await this.server.init(this.handleMessage.bind(this), this.logger);
+    this.logger.log('NATS Startup Service initialized');
+  }
```
There seems to be an issue here which is causing a blocker in my testing/review. The application starts up just fine. But when a message is publish on the `cms` subject it reads the message and reconnects to the NATS server and that is it. 

The code does not seem to be moving forward after that. I have been trying to identify the issue but have not had any luck. Do let me know if this is not what you have experienced and I might be doing something wrong. 

<img width="996" height="241" alt="image" src="https://github.com/user-attachments/assets/1bc6195b-b6aa-417e-8409-0894ce389329" />

mataa019 replied:
Done fixing this

UmairKhan-Paysys reviewed (changes requested):
Ran into some issues. Left comments.

mataa019 committed:
feat:fix nats server and logger service

mataa019 committed:
feat:updating the .env.example

Akku-Rawat committed:
feat(manual-alert): update getUserAlerts api response

- updated CaseType enum and prisma schema

- updated getUserAlerts api to return source and alert type

- updated getUserAlerts filter logic

- fixed failing tests

- removed redundant investigateAlertDto and its tests

Akku-Rawat committed:
refactor(cms): refactored folder structure

mwan9ilwa committed:
merging cms dashboard ui components into manual alert handling

Akku-Rawat committed:
refactor(manual-alert): update convertAlertToCase api request body

- added caseOwnerUserId field to convertAlertToCase dto

- updated convertAlertToCase dto test cases

UmairKhan-Paysys commented on backend/README.md:
````diff
@@ -0,0 +1,531 @@
+    end
+```
+
+## Table of Contents
+
+- [1. Component Overview](#1-component-overview)
+- [2. System Architecture](#2-system-architecture)
+  - [2.1 Authentication Flow](#21-authentication-flow)
+- [3. Configuration](#3-configuration)
+- [5. Running the Service](#5-running-the-service)
+- [6. Testing](#6-testing)
+- [7. Coding Standards](#7-coding-standards)
+- [8. Troubleshooting](#8-troubleshooting)
+- [9. Security](#9-security)
+
+---
+
+## **_1. Component Overview_**
+
+This is a NestJS + TypeScript service for managing financial crime cases and investigations. It includes modules for triage, alert-to-case conversion, tasking, evidence, reporting, auditing, and authentication/authorization (Keycloak via Tazama Auth Service). It uses PostgreSQL via Prisma and supports multi-tenant RBAC.
````
The Readme.md repeats it's contents. Can you please fix this. Thank you.

mataa019 replied:
The repeated content in readme has been removed and the file is now updated

UmairKhan-Paysys commented on backend/prisma/schema.prisma:
```diff
@@ -0,0 +1,152 @@
+  alert_id       String   @id @default(uuid()) @db.Uuid
+  tenant_id      String   @db.Uuid
+  priority       Priority? @default(LOW)
+  alert_type     AlertType?
```
Since we talked about the AI-Model calculating whether or not the alert is a False_positive or a True_Positive. I think there should be a alert_matrix field for the outcome.
```suggestion
  alert_type     AlertType?
  alert_matrix AlertMatrix?
```

mataa019 replied:
The alert_matrix (we are using this name   prediction_outcome ) field has now been added to the alert model in schema.prisma to capture the AI-model outcome.

UmairKhan-Paysys commented on backend/prisma/schema.prisma:
```diff
@@ -0,0 +1,152 @@
+  FRAUD
+  AML
+  FRAUD_AND_AML
+}
```
Enum for the AlertMatrix field. I'm not satisfied with the field name. Feel free to come up with a better one. I appreciate it.
```suggestion
}

enum AlertMatrix {
FALSE_POSITIVE
TRUE_POSITIVE
FALSE_NEGATIVE
TRUE_NEGATIVE
}
```

mataa019 replied:
We added the AlertMatrix (we are using this name prediction_outcome ) enum with the suggested values.

UmairKhan-Paysys commented on backend/prisma/schema.prisma:
```diff
@@ -0,0 +1,152 @@
+  message        String   @db.VarChar(100)
+  alert_data     Json
+  transaction    Json
+  network_map    Json
```
This makes sure the object is saved as a jsonb in postgres
```suggestion
  alert_data     Json  @db.JsonB
  transaction    Json  @db.JsonB
  network_map    Json  @db.JsonB
```

mataa019 replied:
I’ve updated the fields to use @db.JsonB to ensure they are stored as jsonB in Postgres.

UmairKhan-Paysys commented on backend/src/app.module.ts:
```diff
@@ -0,0 +1,33 @@
+  imports: [
+    ConfigModule.forRoot({
+      envFilePath: '.env',
+      isGlobal: true,
```
I recommend adding environment variables validation to make sure all envs are defined during deployment.
```suggestion
      isGlobal: true,
      validate,
```

mataa019 replied:
I’ve added the validate option to ConfigModule.forRoot to ensure all required environment variables are validated during deployment.

UmairKhan-Paysys commented on backend/src/app.module.ts:
```diff
@@ -0,0 +1,33 @@
+import { TokenExpiryInterceptor } from './auth/token-expiry.interceptor';
+import { TriageModule } from './triage/triage.module';
+import { NatsModule } from './nats/nats.module';
+
```
```suggestion
import { validate } from './config/env.validation';

```

UmairKhan-Paysys commented on backend/src/app.module.ts:
For the config validation bit, please refer to the following [repo](https://github.com/UmairKhan-Paysys/case-management-system/tree/manual-alert-handling) 
Please view the `env.validation.ts` in the `src/config` folder

mataa019 replied:
For config validation, we are using the env.validation.ts file in the config folder as recommended.

UmairKhan-Paysys commented on backend/src/auth/auth.controller.ts:
I made some changes here when I was implementing authentication using the `@tazama-lf/auth-lib`
Please check [here](https://github.com/UmairKhan-Paysys/case-management-system/tree/manual-alert-handling)

mataa019 replied:
We have completed making the changes for authentication using @tazama-lf/auth-lib.

UmairKhan-Paysys commented on backend/src/auth/auth.module.ts:
This was also updated. [Link.](https://github.com/UmairKhan-Paysys/case-management-system/tree/manual-alert-handling)

mataa019 replied:
We have completed making the changes for auth module

UmairKhan-Paysys commented on backend/src/auth/jwt.strategy.ts:
This file was completely removed in my [implementation](https://github.com/UmairKhan-Paysys/case-management-system/tree/manual-alert-handling).
Instead there should exists a `tazama-auth.guard.ts` file (with a few others) which is now being used for authorization and utilizes the `@tazama-lf/auth-lib`

UmairKhan-Paysys commented on backend/src/auth/roles.decorator.ts:
Removed as well.

UmairKhan-Paysys commented on backend/src/auth/roles.guard.ts:
Not needed in my [implementation](https://github.com/UmairKhan-Paysys/case-management-system/tree/manual-alert-handling)

UmairKhan-Paysys commented on backend/src/main.ts:
```diff
@@ -0,0 +1,36 @@
+
+  const app = await NestFactory.create(AppModule, {
+    logger,
+  });
```
Your implementation creates another instance of the loggerService here. Might I suggest doing it like:
```suggestion
import { ValidationPipe } from '@nestjs/common';
import 'dotenv/config';

async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  const logger = app.get(LoggerService);
  app.useLogger(logger);
```

Also, please make sure to install and import the `dotenv` package. Otherwise the loggerService might cause issues as NestJS handles configs a bit differently.

UmairKhan-Paysys commented on backend/src/nats/dto/AlertMessageDto.dto.ts:
I updated the AlertMessageDTO and also created a new one for nested `transaction object` validation.
[Link.](https://github.com/UmairKhan-Paysys/case-management-system/tree/manual-alert-handling)

UmairKhan-Paysys commented on backend/src/nats/nats.startup.ts:
Please refer to [this.](https://github.com/UmairKhan-Paysys/case-management-system/tree/manual-alert-handling)

UmairKhan-Paysys commented on backend/src/triage/dto/submit-alert.dto.ts:
Please refer [here.](https://github.com/UmairKhan-Paysys/case-management-system/tree/manual-alert-handling)

UmairKhan-Paysys commented on backend/src/triage/dto/update-alert.dto.ts:
```diff
@@ -0,0 +1,16 @@
+
+  @IsOptional()
+  @IsEnum(AlertType)
+  alertType?: AlertType;
```
Addition to the manual update.

```suggestion
  alertType?: AlertType;
  
  @IsOptional()
  @IsEnum(AlertMatrix)
  alert_matrix?: AlertMatrix;
```

UmairKhan-Paysys commented on backend/src/triage/dto/update-alert.dto.ts:
```diff
@@ -0,0 +1,16 @@
+import { Priority, AlertType } from '@prisma/client';
```
```suggestion
import { Priority, AlertType, AlertMatrix } from '@prisma/client';
```

UmairKhan-Paysys commented on backend/src/triage/triage.controller.ts:
```diff
@@ -0,0 +1,109 @@
+import { RolesGuard } from '../auth/roles.guard';
+
+@Controller('api/v1/triage/alerts')
+@UseGuards(AuthGuard('jwt'), RolesGuard)
```
Utilizing the updated implementation here.

```suggestion
@UseGuards(TazamaAuthGuard)
```

UmairKhan-Paysys commented on backend/src/triage/triage.controller.ts:
There are more changes in this file. I'll simply add the link to the fork in the main review comment.

UmairKhan-Paysys commented on backend/src/triage/triage.service.ts:
Made changes here as well.

UmairKhan-Paysys reviewed (changes requested):
I appreciate the frontend being pushed to this PR. Please keep this as a `feature-consolidation` branch for now. For the requested changes on how to implement authorization and NATS. Please refer to [this.](https://github.com/UmairKhan-Paysys/case-management-system/tree/manual-alert-handling). Thank you.

Akku-Rawat committed:
refactor(manual-alert): add PredictionOutcome enum and update alert schema

- Added PredictionOutcome enum

- Changed alert schema JSON to JSONB and added prediction_outcome field

- Updated SubmitAlertDto and service logic to include prediction_outcome

- Removed duplicate section from README

- Removed filterOptions endpoint and related service logic

mwan9ilwa committed:
fix: removed hardcoded risk strings and replace them with parsed values from alert.alert_data

mataa019 committed:
feat:fix auth module and nats files updating logger in main

mwan9ilwa committed:
fix: Used provided props to populate the filter dropdowns — fallback to static defaults when props are empty or undefined

mwan9ilwa committed:
fix: Removed the now-deprecated onSearch prop passed from AlertsDashboard.tsx to AlertsSearchAndFilters.tsx

mwan9ilwa committed:
fix: beautified the transaction data response

mataa019 committed:
updated the .env.example

mataa019 committed:
updated the .env.example

mataa019 committed:
updated the .env.example

Akku-Rawat committed:
refactor(manual-alert): implemented feedback

Akku-Rawat committed:
refactor: use authetAuthenticatedRequest type in controller

UmairKhan-Paysys commented:
<img width="1070" height="705" alt="image" src="https://github.com/user-attachments/assets/0d866632-5c44-40c8-b1f2-126289f9a51a" />
Facing issue getting the frontend to work. Even resolving these conflicts would not resolve all the errors preventing me to make a build and start up the client.

mwan9ilwa commented:
> <img alt="image" width="1070" height="705" src="https://private-user-images.githubusercontent.com/185335318/483315409-0d866632-5c44-40c8-b1f2-126289f9a51a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY0NDcyNzAsIm5iZiI6MTc1NjQ0Njk3MCwicGF0aCI6Ii8xODUzMzUzMTgvNDgzMzE1NDA5LTBkODY2NjMyLTVjNDQtNDBjOC1iMWYyLTEyNjI4OWY5YTUxYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyOVQwNTU2MTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NTcwMjY3MzNlZDQwMThhNjg0NjgyZTNkNzdhNzdmNTkyYjM3YzY1MDI2MmY5MDA4NzM2Mzc3OGE4ZTdmODlhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BILqvOd7gaFJb6ZiXEdGcsg6TPSwJiXItXtzOGIkhLo"> Facing issue getting the frontend to work. Even resolving these conflicts would not resolve all the errors preventing me to make a build and start up the client.

When last did you pull the latest code? My team are facing no issue at all pulling the latest code.

Akku-Rawat committed:
refactor(ai-triage): implemented feedback

Akku-Rawat committed:
refactor(ai-triage): removed priority from placehholder ai

dev-Gerald committed:
unit tests

dev-Gerald committed:
unit tests

dev-Gerald committed:
Revert "unit tests"

This reverts commit 860a794ebfb5ca2074e7b54f703cbe89e5e2f651.

dev-Gerald committed:
Revert "unit tests"

This reverts commit 860a794ebfb5ca2074e7b54f703cbe89e5e2f651.

UmairKhan-Paysys commented:
> > <img alt="image" width="1070" height="705" src="https://private-user-images.githubusercontent.com/185335318/483315409-0d866632-5c44-40c8-b1f2-126289f9a51a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY0NDcyNzAsIm5iZiI6MTc1NjQ0Njk3MCwicGF0aCI6Ii8xODUzMzUzMTgvNDgzMzE1NDA5LTBkODY2NjMyLTVjNDQtNDBjOC1iMWYyLTEyNjI4OWY5YTUxYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyOVQwNTU2MTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NTcwMjY3MzNlZDQwMThhNjg0NjgyZTNkNzdhNzdmNTkyYjM3YzY1MDI2MmY5MDA4NzM2Mzc3OGE4ZTdmODlhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BILqvOd7gaFJb6ZiXEdGcsg6TPSwJiXItXtzOGIkhLo"> Facing issue getting the frontend to work. Even resolving these conflicts would not resolve all the errors preventing me to make a build and start up the client.
> 
> When last did you pull the latest code? My team are facing no issue at all pulling the latest code.

Did it after your last commit. I'm not sure why I'm getting merge conflicts

mwan9ilwa committed:
feat: implement manual alert handling functionality

- Add AlertsDetailModal with alert details display
- Add CloseAlertModal for alert closure workflow
- Add UpdateAlertModal for alert updates
- Add DashboardHeader and ResultsSummary components
- Add useAlerts and useAlertOperations hooks
- Update AlertsDashboard with new functionality
- Update triage service and types
- Remove ConvertToCaseModal component

mwan9ilwa committed:
Merge remote-tracking branch 'origin/manual-alert-handling' into manual-alert-handling

Resolved conflicts by keeping local changes for:
- frontend/src/pages/AlertsDashboard.tsx
- frontend/src/components/common/CloseAlertModal.tsx

Integrated upstream changes including:
- Backend authentication refactoring
- New case and task management modules
- Updated triage service functionality

mwan9ilwa committed:
fix: resolve merge conflict markers in UpdateAlertModal and AlertsSearchAndFilters

- Cleaned up Git conflict markers that were causing build errors
- Updated AlertsSearchAndFilters to use prop-based filter options
- Fixed function parameter formatting in UpdateAlertModal

mwan9ilwa committed:
Removed conficted and corrupt directories

mwan9ilwa committed:
setup testing for the frontend

dev-Gerald committed:
sla priority

dev-Gerald committed:
sla priority

mwan9ilwa committed:
updated client side filtering

mwan9ilwa committed:
Merge branch 'manual-alert-handling' of https://github.com/tazama-lf/case-management-system into manual-alert-handling

mwan9ilwa commented:
> > > <img alt="image" width="1070" height="705" src="https://private-user-images.githubusercontent.com/185335318/483315409-0d866632-5c44-40c8-b1f2-126289f9a51a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY0NDcyNzAsIm5iZiI6MTc1NjQ0Njk3MCwicGF0aCI6Ii8xODUzMzUzMTgvNDgzMzE1NDA5LTBkODY2NjMyLTVjNDQtNDBjOC1iMWYyLTEyNjI4OWY5YTUxYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyOVQwNTU2MTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NTcwMjY3MzNlZDQwMThhNjg0NjgyZTNkNzdhNzdmNTkyYjM3YzY1MDI2MmY5MDA4NzM2Mzc3OGE4ZTdmODlhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BILqvOd7gaFJb6ZiXEdGcsg6TPSwJiXItXtzOGIkhLo"> Facing issue getting the frontend to work. Even resolving these conflicts would not resolve all the errors preventing me to make a build and start up the client.
> > 
> > 
> > When last did you pull the latest code? My team are facing no issue at all pulling the latest code.
> 
> Did it after your last commit. I'm not sure why I'm getting merge conflicts


Try now @UmairKhan-Paysys I've resolved any conflicts.
